### ISSUE TITLE
bucket_by_sequence_length does not dequeue all items given to it from another queue

### ISSUE THREAD

**Issue by nOkuda** [tensorflow/tensorflow#8182]
I've got some SequenceExamples gzipped in a file that I read with TFRecordReader.  I then enqueue those serialized SequenceExamples onto a RandomShuffleQueue.  Then, I dequeue the RandomShuffleQueue, get a number I need from the serialized SequenceExample, and pass both the number and the SequenceExample into bucket_by_sequence_length.

Since there are 100 items stored in the file, I expected to get 20 batches of 5.  Instead, I'm getting 17 batches of 5, with many of the smallest input_length items missing.

My guess is that the top queue on batch_by_sequence_length is looking at its bucket queues, finding that only very small and very large inputs remain, and so refuses to put them together for the final batch(es).  I was hoping that by setting `allow_smaller_final_batch` to `True`, I could avoid that problem, but I guess not.

### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?

GitHub issue #5609 (the discussion at the very bottom of this thread seems to be the only place on the web [as of 7 Mar 2017] with a discussion on how to actually use bucket_by_sequence_length)

### Environment info
Operating System:  Fedora 24

Installed version of CUDA and cuDNN:  8.0.44, 5.1.5
(please attach the output of `ls -l /path/to/cuda/lib/libcud*`):
-rw-r--r-- 1 root root   558720 Nov 24 09:02 libcudadevrt.a
lrwxrwxrwx 1 root root       16 Nov 24 09:02 libcudart.so -> libcudart.so.8.0
lrwxrwxrwx 1 root root       19 Nov 24 09:02 libcudart.so.8.0 -> libcudart.so.8.0.44
-rwxr-xr-x 1 root root   415432 Nov 24 09:02 libcudart.so.8.0.44
-rw-r--r-- 1 root root   775162 Nov 24 09:02 libcudart_static.a
-rwxr-xr-x 1 root root 79337624 Nov 24 09:05 libcudnn.so
-rwxr-xr-x 1 root root 79337624 Nov 24 09:05 libcudnn.so.5
-rwxr-xr-x 1 root root 79337624 Nov 24 09:05 libcudnn.so.5.1.5
-rw-r--r-- 1 root root 69756172 Nov 24 09:05 libcudnn_static.a

If installed from binary pip package, provide:

1. A link to the pip package you installed:  https://ci.tensorflow.org/view/Nightly/job/nightly-matrix-linux-gpu/TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=PIP,TF_BUILD_PYTHON_VERSION=PYTHON3,label=gpu-linux/lastSuccessfulBuild/artifact/pip_test/whl/tensorflow_gpu-1.0.0-cp35-cp35m-linux_x86_64.whl <- installed on 3 Mar 2017
2. The output from `python3 -c "import tensorflow; print(tensorflow.__version__)"`.  1.0.0

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)

Save the following code as `test_bucket.py`, download [other.gz](https://github.com/tensorflow/tensorflow/files/826517/other.gz) into the same directory, then run 
```python3 test_bucket.py --intput_file other.gz --run_type batches```

```Python
import argparse
import tensorflow as tf


LAST_RUNNER = 'batch_dequeue'
CONTEXT_FEATURES = {
    'aas_length': tf.FixedLenFeature([], dtype=tf.int64),
    'funcs_length': tf.FixedLenFeature([], dtype=tf.int64)}
SEQUENCE_FEATURES = {
    'aas': tf.FixedLenSequenceFeature([], dtype=tf.int64),
    'funcs': tf.FixedLenSequenceFeature([], dtype=tf.int64)}


def read_aas_length(serialized):
    """Read aas_length"""
    context_parsed, _ = tf.parse_single_sequence_example(
        serialized=serialized,
        context_features=CONTEXT_FEATURES,
        sequence_features=SEQUENCE_FEATURES)
    return context_parsed['aas_length']


def main(args):
    with tf.Session() as sess:
        # filename_queue is FIFOQueue
        filename_queue = tf.train.string_input_producer(
            [args.input_file],
            num_epochs=1,
            name='tfrecord_filename_queue')
        reader = tf.TFRecordReader(
            name='tfrecord_reader',
            options=tf.python_io.TFRecordOptions(
                tf.python_io.TFRecordCompressionType.GZIP))

        _, next_raw = reader.read(
            filename_queue,
            name='read_records')
        random_raws = tf.RandomShuffleQueue(
            capacity=50,
            min_after_dequeue=0,
            dtypes=tf.string,
            # http://stackoverflow.com/questions/42119238/tensorflow-fifoqueues-dequeuemany-and-dequeueupto-require-the-components-to-ha
            shapes=[()],
            name='randomize_records')
        enqueue_random_raws = random_raws.enqueue(next_raw)

        serialized_example_dq = random_raws.dequeue()
        aas_length = read_aas_length(serialized_example_dq)
        if args.run_type == 'batches':
            batch_max_lens, batches = \
                tf.contrib.training.bucket_by_sequence_length(
                    tf.to_int32(aas_length),
                    [serialized_example_dq],
                    5,
                    [100, 200, 300, 400, 500, 1000],
                    allow_smaller_final_batch=True,
                    name=LAST_RUNNER)

        init = tf.global_variables_initializer()
        sess.run(init)
        # necessary when num_epochs in string_input_producer is not None
        sess.run(tf.local_variables_initializer())

        random_records_runner = tf.train.QueueRunner(
            random_raws,
            [enqueue_random_raws] * 4)
        coord = tf.train.Coordinator()
        tf.train.add_queue_runner(random_records_runner)
        threads = tf.train.start_queue_runners(coord=coord)
        working = True
        i = 0

        if args.run_type == 'batches':
            fetch = {
                'batch_max_lens': batch_max_lens,
                'batches': batches}

        while working:
            try:
                i += 1
                if args.run_type == 'batches':
                    fetched = sess.run(fetch)
                    print(i, fetched['batch_max_lens'])
                else:
                    print(i, sess.run(aas_length))
            except tf.errors.OutOfRangeError as err:
                coord.request_stop()
                working = False

        coord.join(threads)


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument(
        '--input_file',
        type=str,
        required=True,
        help='Serialized file path... ?')
    parser.add_argument(
        '--run_type',
        type=str,
        default='',
        help='Choose "batches" to run in batches')
    args = parser.parse_args()
    main(args)
```

### What other attempted solutions have you tried?

I've tried collecting relevant data in the while loop and then feeding them into a placeholder for batch_by_sequence_length.  That yielded an error message about expecting a string but getting a _Message, or something to that effect.

I've also looked at the node_def.name of the OutOfRangeError.  It only ever came up with the top queue created by bucket_by_sequence_length.

Somewhat related, calling bucket_by_sequence_length but not fetching its outputs (so taking the function call out of the if block and running the code without the `--run_type` flag) still caused the list of input lengths to stop at 99 (not 100, as I would expect).

I've also observed that the missing smaller numbers will show up if `num_epochs` on the TFRecordReader is set to 2, although they only appear once.

There were other failed attempts, but I don't recall their details.

### Logs or other output that would be helpful
[output_batches.txt](https://github.com/tensorflow/tensorflow/files/826529/output_batches.txt):  Example output of not getting back enough batches.
[output_list.txt](https://github.com/tensorflow/tensorflow/files/826530/output_list.txt):  Example output of printing out all lengths (note that some lengths are less than 100, but there are no lengths in the batches less than 100).

**Comment by prb12** [comment]
@joel-shor @ebrevdo  Could one of you please comment on whether this behavior is expected?

**Comment by joel-shor** [comment]
@nOkuda would it be possible for you to shorten your example to a minimal amount of code? That would make it easier for me to investigate this issue. If the problem is with `training.bucket_by_sequence_length` not batching short and long examples together, you should be able to create a relatively small example demonstrating this.

I'll still investigate either way, but it might take longer for me to be helpful.

**Comment by nOkuda** [comment]
@joel-shor Sorry about not making a better bit of code.  I was originally thinking that the problem was arising from reading tensorflow records or chaining together multiple queues.  But the new minimal working example definitely supports the hypothesis you chose to point out.

Updated minimal working example:
```Python
import argparse
import tensorflow as tf


LAST_RUNNER = 'batch_dequeue'


def main():
    with tf.Session() as sess:
        examples_queue = tf.train.string_input_producer(
            ['1500', '1700', '10', '5', '1'],
            num_epochs=1,
            name='examples_queue')
        example_dq = examples_queue.dequeue()

        batch_max_lens, batches = \
            tf.contrib.training.bucket_by_sequence_length(
                tf.string_to_number(example_dq, out_type=tf.int32),
                [example_dq],
                5,
                [100, 200, 300, 400, 500, 1000],
                allow_smaller_final_batch=True,
                name=LAST_RUNNER)

        init = tf.global_variables_initializer()
        sess.run(init)
        sess.run(tf.local_variables_initializer())

        coord = tf.train.Coordinator()
        threads = tf.train.start_queue_runners(coord=coord)
        working = True
        i = 0

        fetch = {
            'batch_max_lens': batch_max_lens,
            'batches': batches}

        while working:
            try:
                i += 1
                fetched = sess.run(fetch)
                print(i, fetched['batch_max_lens'])
            except tf.errors.OutOfRangeError as err:
                coord.request_stop()
                working = False

        coord.join(threads)


if __name__ == '__main__':
    main()
```

**Comment by joel-shor** [comment]
What bug are you trying to demonstrate with this code?

I see two issues with your example:
1) You need to call `coord.request_stop()`
2) With `string_input_producer` only having one epoch, you will always get an `OutOfRange` exception before any of your `bucket_by_sequence_length` queues are full with 5 examples

Would you mind re-posting a minimal working example with the bug you are trying to highlight? Are you positive that `bucket_by_sequence_length` behaves the way you think it does?

**Comment by nOkuda** [comment]
I'm expecting `bucket_by_sequence_length` to dequeue all buckets once it's reached the end of the queue that's feeding into it.  Instead, it does not.

If what I'm expecting is not intended behavior, how do I dequeue the buckets once the feeding queue is empty?

To address your two points:  I am calling `coord.request_stop()` when I catch the `OutOfRangeError` exception.  Also, the `OutOfRangeError` exception is coming from the top queue built by `bucket_by_sequence_length`, not from the `string_input_producer`.  To see the `OutOfRangeError` behavior I have described, add a `print(err.node_def.name)` in the `except` block.

I accept that the top queue of `bucket_by_sequence_length` may be passing up the `OutOfRangeError` from `string_input_producer`; but if that is the case, then I would expect `err.node_def.name` to be related to the dequeue from the `string_input_producer`.  Instead, the output name is `batch_dequeue/bucket/dequeue_top`.

I'm also recalling reading that `OutOfRangeError`s do not come from queues dequeueing into another queue.  I can't remember where I read that, though.  Something about threads blocking when there's nothing left to dequeue.  Maybe it was in the [Threading and Queueing](https://www.tensorflow.org/programmers_guide/threading_and_queues) instructions?  If doesn't work like that, I have no idea how `bucket_by_sequence_length` would work in the lazy manner it purports to in the [documentation](https://www.tensorflow.org/versions/master/api_docs/python/contrib.training/bucketing#bucket_by_sequence_length).

**Comment by yifeif** [comment]
@joel-shor could you take another look? Thanks!

**Comment by bdaskalov** [comment]
@nOkuda @yifeif Did you find a solution? I ran into the same problem and I have some idea why it happens but it won't be trivial to fix.

**Comment by nOkuda** [comment]
@bdaskalov With respect to the project I was using this feature in, my solution was simply to run so many epochs that I wouldn't possibly run into the bug (i.e., training took so long that I never reached the end---this was fine, since I was saving out the information I needed after each epoch; when it was time to wrap up the project, I simply took the latest information and used that).

As for the bug I reported in this issue, I have not found a solution.

**Comment by mschulkind** [comment]
I think fixing the problem is fairly easy, although since `QueueRunner` isn't really designed for this, it's slightly messy, but better than the current situation. The problem right now is that the `top_queue` is closed as soon as the input to `bucket_sequence_by_length` runs dry, and then nobody ever closes the `bucket_queues`, but it wouldn't even matter, since by then, the `top_queue` is already closed.

This seems to fix the problem:
```
    queue_runner.add_queue_runner(
      queue_runner.QueueRunner(
          bucket_queues[0], enqueues_to_top,
          close_op=top_queue.close(),
          queue_closed_exception_types=(errors.OutOfRangeError,
                                        errors.CancelledError)))
    queue_runner.add_queue_runner(
        queue_runner.QueueRunner(
            top_queue,
            bucket_enqueue_ops,
            close_op=control_flow_ops.group(*[q.close() for q in bucket_queues]),
            queue_closed_exception_types=(errors.OutOfRangeError,
                                          errors.CancelledError)))
```

Basically, instead of closing `top_queue` when the overall input runs dry, this closes all of the bucket queues, so they then start feeding their last batches. It also uses only one `QueueRunner` for the `enqueues_to_top`, so that after all of them individually run dry, the `top_queue` is finally closed.

**Comment by mschulkind** [comment]
The PR above should fix this.

**Comment by ebrevdo** [comment]
I like it!  Will review this week.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by yifeif** [comment]
The corresponding PR has been merged internally and should be pushed out soon. Thanks @mschulkind!

**Comment by tensorflowbutler** [comment]
Nagging Assigneee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
